### PR TITLE
Fix broken IPv6 addresses handling

### PR DESCRIPTION
**Pull request by cmdrclueless**
See Issue #19784 and closed pull #19786 for reference.

### CODE REVIEW

**Comment by al2o3cr** [comment]
Question: why is `ActionDispatch::Http::URL` splitting out protocol / host / port instead of using the built-in functionality in stdlib's `URI`? I suspect it might be for performance reasons, but digging through history all the way back to 3.0 didn't show any discussion. Using the stdlib version would have prevented this bug.

File: actionpack/test/dispatch/routing/ipv6_redirect_test.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
+require 'abstract_unit'
+
+class IPv6IntegrationTest < ActionDispatch::IntegrationTest
```
Comment (rafaelfranca) [inline actionpack/test/dispatch/routing/ipv6_redirect_test.rb]: Can't we just include this test in n existent test file?

**Comment by rafaelfranca** [comment]
@al2o3cr good question. I can't think in any special reason.

**Comment by joallard** [comment]
`TestRoutingMapper` is ~4500 lines long, maybe this would help split it up a little? Otherwise, is there some other work needed? Seems to me that Rails' IPv6 support is contingent on this patch.

**Comment by robin850** [comment]
@cmdrclueless : Any chance to address Rafael's feedback please ? Thanks for the patch so far!

**Comment by cmdrclueless** [comment]
@robin850 to which comment are you referring? Assuming it's about the rational of putting IPv6 testing in a separate file... well @joallard more or less hit the nail on the head. It seemed reasonable to put the actual test in a separate file to limit the scope to the actual problem I was addressing.

**Comment by rafaelfranca** [comment]
Backported in e7c68d0a4d5c84db107e43c9cc4f7baee2ed00b1

**Pull request merged**